Bogus error "lambda.param.type.incompatible"

The following program can be compiled by javac. If, however, the NullnessChecker is used, one gets 

_[9,33] error: [lambda.param.type.incompatible] incompatible parameter types in lambda expression._
```
package bar;

import java.util.stream.Stream;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;

public class Bar {
	public static Stream<? extends ZipEntry> entries(ZipFile zipFile) {
		return zipFile.stream().filter(entry -> !entry.isDirectory() && entry.getName().endsWith(".xml"));
	}
}
```

No error is issued if you write
`(ZipEntry entry) -> !entry.isDirectory() && ...`

Here's the full error message:
````
javacheck -processor nullness java8/lambda/Issue870.java -AprintVerboseGenerics -AprintAllQualifiers
java8/lambda/Issue870.java:12: error: [lambda.param.type.incompatible] incompatible parameter types in lambda expression.
                .filter(entry -> !entry.isDirectory() && entry.getName().endsWith(".xml"));
                        ^
  found   : ?[ extends @UnknownKeyFor ZipEntry super @KeyForBottom Void]
  required: capture#47 of ? extends java.util.zip.ZipEntry[ extends @UnknownKeyFor ZipEntry super @KeyForBottom Void]
1 error
````

--- a/framework/src/org/checkerframework/framework/type/DefaultTypeHierarchy.java
+++ b/framework/src/org/checkerframework/framework/type/DefaultTypeHierarchy.java
@@ -12,6 +12,8 @@ import java.util.Set;
 import javax.lang.model.element.AnnotationMirror;
 import javax.lang.model.element.TypeElement;
 import javax.lang.model.type.TypeKind;
+import javax.lang.model.type.TypeMirror;
+import javax.lang.model.type.TypeVariable;
 import javax.lang.model.util.Elements;
 import javax.lang.model.util.Types;
 import org.checkerframework.common.basetype.BaseTypeChecker;
@@ -30,6 +32,7 @@ import org.checkerframework.framework.util.AtmCombo;
 import org.checkerframework.framework.util.PluginUtil;
 import org.checkerframework.framework.util.TypeArgumentMapper;
 import org.checkerframework.javacutil.ErrorReporter;
+import org.checkerframework.javacutil.InternalUtils;
 import org.checkerframework.javacutil.Pair;
 import org.checkerframework.javacutil.TypesUtils;
 
@@ -919,30 +922,6 @@ public class DefaultTypeHierarchy extends AbstractAtmComboVisitor<Boolean, Visit
     @Override
     public Boolean visitWildcard_Wildcard(
             AnnotatedWildcardType subtype, AnnotatedWildcardType supertype, VisitHistory visited) {
-        // This method is called at a method invocation where a type variable in the method
-        // declaration is substituted with a wildcard.
-        // For example:
-        // <T> void method(Gen<T> t) {}
-        // Gen<?> x;
-        // method(x); // this method is called when checking this method call
-
-        if (subtype.getUnderlyingType() == supertype.getUnderlyingType()) {
-            boolean subtypeHasAnno = subtype.getAnnotationInHierarchy(currentTop) != null;
-            boolean supertypeHasAnno = supertype.getAnnotationInHierarchy(currentTop) != null;
-
-            if (subtypeHasAnno && supertypeHasAnno) {
-                // if both have primary annotations then just check the primary annotations
-                // as the bounds are the same
-                return isPrimarySubtype(subtype, supertype, true);
-
-            } else if (!subtypeHasAnno
-                    && !supertypeHasAnno
-                    && areEqualInHierarchy(subtype, supertype, currentTop)) {
-                // TODO: wildcard capture conversion
-                // Two unannotated uses of reference-equal wildcard types are the same type
-                return true;
-            }
-        }
         return visitWildcardSubtype(subtype, supertype, visited);
     }
 
@@ -1006,6 +985,40 @@ public class DefaultTypeHierarchy extends AbstractAtmComboVisitor<Boolean, Visit
 
     protected boolean visitWildcardSubtype(
             AnnotatedWildcardType subtype, AnnotatedTypeMirror supertype, VisitHistory visited) {
+        TypeMirror superTypeMirror = supertype.getUnderlyingType();
+        if (supertype.getKind() == TypeKind.TYPEVAR) {
+            TypeVariable atv = (TypeVariable) supertype.getUnderlyingType();
+            if (InternalUtils.isCaptured(atv)) {
+                superTypeMirror = InternalUtils.getCapturedWildcard(atv);
+            }
+        }
+
+        if (subtype.getUnderlyingType() == superTypeMirror) {
+            // This can happen at a method invocation where a type variable in the method
+            // declaration is substituted with a wildcard.
+            // For example:
+            // <T> void method(Gen<T> t) {}
+            // Gen<?> x;
+            // method(x); // this method is called when checking this method call
+            // And also when checking lambdas
+
+            boolean subtypeHasAnno = subtype.getAnnotationInHierarchy(currentTop) != null;
+            boolean supertypeHasAnno = supertype.getAnnotationInHierarchy(currentTop) != null;
+
+            if (subtypeHasAnno && supertypeHasAnno) {
+                // if both have primary annotations then just check the primary annotations
+                // as the bounds are the same
+                return isPrimarySubtype(subtype, supertype, true);
+
+            } else if (!subtypeHasAnno
+                    && !supertypeHasAnno
+                    && areEqualInHierarchy(subtype, supertype, currentTop)) {
+                // TODO: wildcard capture conversion
+                // Two unannotated uses of reference-equal wildcard types are the same type
+                return true;
+            }
+        }
+
         return isSubtype(subtype.getExtendsBound(), supertype, visited);
     }
 

--- a/javacutil/src/org/checkerframework/javacutil/InternalUtils.java
+++ b/javacutil/src/org/checkerframework/javacutil/InternalUtils.java
@@ -16,6 +16,7 @@ import com.sun.tools.javac.code.Symbol;
 import com.sun.tools.javac.code.Symbol.TypeSymbol;
 import com.sun.tools.javac.code.Type;
 import com.sun.tools.javac.code.Type.AnnotatedType;
+import com.sun.tools.javac.code.Type.CapturedType;
 import com.sun.tools.javac.code.Types;
 import com.sun.tools.javac.processing.JavacProcessingEnvironment;
 import com.sun.tools.javac.tree.JCTree;
@@ -244,6 +245,19 @@ public class InternalUtils {
             return ((Type.TypeVar) ((Type.AnnotatedType) typeVar).unannotatedType()).isCaptured();
         }
         return ((Type.TypeVar) typeVar).isCaptured();
+    }
+
+    /**
+     * If typeVar is a captured wildcard, returns that wildcard; otherwise returns null.
+     */
+    public static WildcardType getCapturedWildcard(TypeVariable typeVar) {
+        if (isCaptured(typeVar)) {
+            if (typeVar instanceof AnnotatedType) {
+                return ((CapturedType) ((AnnotatedType) typeVar).unannotatedType()).wildcard;
+            }
+            return ((CapturedType) typeVar).wildcard;
+        }
+        return null;
     }
 
     /**

--- a/javacutil/src/org/checkerframework/javacutil/InternalUtils.java
+++ b/javacutil/src/org/checkerframework/javacutil/InternalUtils.java
@@ -241,10 +241,7 @@ public class InternalUtils {
      * Returns whether a TypeVariable represents a captured type.
      */
     public static boolean isCaptured(TypeVariable typeVar) {
-        if (typeVar instanceof AnnotatedType) {
-            return ((Type.TypeVar) ((Type.AnnotatedType) typeVar).unannotatedType()).isCaptured();
-        }
-        return ((Type.TypeVar) typeVar).isCaptured();
+        return ((Type.TypeVar) ((Type.TypeVar) typeVar).unannotatedType()).isCaptured();
     }
 
     /**
@@ -252,10 +249,7 @@ public class InternalUtils {
      */
     public static WildcardType getCapturedWildcard(TypeVariable typeVar) {
         if (isCaptured(typeVar)) {
-            if (typeVar instanceof AnnotatedType) {
-                return ((CapturedType) ((AnnotatedType) typeVar).unannotatedType()).wildcard;
-            }
-            return ((CapturedType) typeVar).wildcard;
+            return ((CapturedType) ((Type.TypeVar) typeVar).unannotatedType()).wildcard;
         }
         return null;
     }

